### ISSUE TITLE
This line may casue some false negative samples

### ISSUE THREAD

**Issue by lanzhuzhu** [fractalego/fewrel_zero_shot#1]
Thanks for your opening source. This is excellent work. I looked carefully and found something that could be improved.

As _all_relations_ is read from the file "all_train_relations.txt", which contains 699 labels for each relation, the  _candidates_ still contains the labels for the correct relation after the [`remove']([https://github.com/fractalego/fewrel_zero_shot/blob/5cddc066d65fbc23de29f16f7fc61bd23bcf8b56/rel_extract/aux_fewrel.py#L132) operation.  Thus, the sampled candidate negative labels have a relatively low probability of belonging to the correct relation. According to my computation, this probability is close to 14.02%. 

`all_relations =  list(set(all_relations))' could fix this potential risk.

**Comment by fractalego** [comment]
Hi lanzhuzhu,
Thank you for your message! It's really important for me to get feedback.

Regarding the duplicates on all_relations: I left it this way because I noticed that training with the duplicates would add 1-2 percent points to the validation score. Your change is consistent with the code but then the model would perform slightly worse.

Effectively, the "correct relations" that should not be there are noise, and noise in the training data acts as a regularizer which avoids overfitting. This is not uncommon with neural networks: you can check out for example "Regularizing Deep Neural Networks by Noise: Its Interpretation and Optimization" by Noh et al.

I hope to write a follow-up paper where this procedure is made more clear. I found the improvement by sheer chance, the paper deadline was upon me and decided to leave it there.

I should have taken out the lines 
```python
try:
    candidates.remove(relation)      
except:
    pass
``` 
but the training had already been done and preferred to push the code as it was.

**Comment by lanzhuzhu** [comment]
Hi, Fractalego. Thanks for your explanation. It cleared away my confusion.
It is interesting to find that such noise brings performance improvement.   It would be great that your follow-up paper could provide more fundamentals underlying this phenomenon.

**Comment by fractalego** [comment]
Thank you again! I hope to make the "noising" system more clear in the future. I found this improvement by sheer chance and it took me a while to understand why it was better this way. 
The "false negative" line was not easy to find! I am humbled you took the time to read this code.
